File: php/src/LombokAlgoritma/Geometry/Bezier.php
<?php

// LombokAlgoritma — Bézier curve evaluation, de Casteljau (SPEC §13.2)
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok

declare(strict_types=1);

namespace LombokAlgoritma\Geometry;

use LombokAlgoritma\AlgoException;

/** de Casteljau evaluation, O(n²). */
final class Bezier
{
    /**
     * Point at parameter t: each level replaces pᵢ with pᵢ·s + pᵢ₊₁·t, s = 1 − t.
     *
     * @param list<Point> $controlPoints
     * @throws AlgoException EMPTY_INPUT with no control points
     */
    public static function at(array $controlPoints, int|float $t): Point
    {
        if ($controlPoints === []) {
            throw AlgoException::emptyInput('bezier (need ≥ 1 control point)');
        }
        $pts = $controlPoints;
        $s = 1 - $t;
        while (count($pts) > 1) {
            $next = [];
            for ($i = 0; $i + 1 < count($pts); $i++) {
                $p = $pts[$i];
                $q = $pts[$i + 1];
                $next[] = new Point($p->x * $s + $q->x * $t, $p->y * $s + $q->y * $t);
            }
            $pts = $next;
        }
        return $pts[0];
    }
}

File: rust/lombokalgoritma/src/geometry/bezier.rs
// LombokAlgoritma — Bézier curve evaluation, de Casteljau (SPEC §13.2)
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
use super::Point2D;
use crate::{Error, Result};
use alloc::vec::Vec;

/// Point of the Bézier curve with the given control points at parameter `t` (de Casteljau):
/// each level replaces `pᵢ` by `pᵢ·s + pᵢ₊₁·t` with `s = 1 − t`. O(n²).
///
/// # Errors
/// [`Error::EmptyInput`] without control points.
pub fn bezier(control_points: &[Point2D], t: f64) -> Result<Point2D> {
    if control_points.is_empty() {
        return Err(Error::EmptyInput);
    }
    let s = 1.0 - t;
    let mut pts: Vec<Point2D> = control_points.to_vec();
    while pts.len() > 1 {
        pts = pts
            .windows(2)
            .map(|w| Point2D::new(w[0].x * s + w[1].x * t, w[0].y * s + w[1].y * t))
            .collect();
    }
    Ok(pts[0])
}
